### ISSUE TITLE
Improve exception handling in RegexParser

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/corefx#28508]
Related to https://github.com/dotnet/corefx/issues/27618

Currently RegexParser always throws ArgumentExceptions with a custom text when something's going wrong.  Do differentiate error cases (in tests) and make other uses cases like TryParse or Validate possible we should refactor the current exception handling. I'm thinking about the following options:

a) Introduce a RegexParseException which inherits from ArgumentException to not break existing code that contains an enum status code that tells the parse error code. From this we can then build the existing error messages. This options doesn't support TryParse / Validate scenarios.
b) Return a RegexParseEnum that tells what went wrong. Ultimately this will be wrapped inside an ArgumentException to not break existing code. TryParse / Validate scenarios can either return this enum or return the error message from it.
c) Keep everything as it is and only tests if ArgumentExceptions are thrown during negative tests.

cc @stephentoub @danmosemsft @CyrusNajmabadi

**Comment by danmosemsft** [comment]
For (b) how would you wrap the enum - put it in the message? Stuff it into Data field?

In customer scenarios (as opposed to unit tests) how interesting is this detailed error data? If not much I'd be inclined to stuff it into Data and not add yet another S.T.RE type.

**Comment by ViktorHofer** [comment]
With b) we don't need to stuff the enum into the ArgumentException, we only need the enum for the string message construction. I agree with not introducing a new type if it doesn't bring any value to the end user. I will look into how feasible b is.

**Comment by danmosemsft** [comment]
I recall the System.Xml team way back declining to add error codes into their XmlException because they noted that refactoring their code (eg changing their parser) could easily change the code, which would have the effect of a breaking change. I don't know whether that is as important as it was perceived to be back then. I think the key question to ask is what the user actually needs. For our own unit tests, scanning the exception message for a non localized keyword (not matching the whole string) is reasonable.

**Comment by ViktorHofer** [comment]
Adding the error code (int or enum) to the existing ArgumentException is easily doable and I don't get how this could be interpreted as a breaking change or regress something. Currently all exceptions in RegexParser.cs are invoked by a MakeException helper method. We would simply change its parameter from a string to an int/enum. The translation to the corresponding string then happens in the MakeException helper function (or in a custom type if we would introduce one).

> For our own unit tests, scanning the exception message for a non localized keyword (not matching the whole string) is reasonable.

Messy... please not...

**Comment by CyrusNajmabadi** [comment]
Sorry... i'm not sure what problem this is trying to solve.  Could you clarify?

**Comment by ViktorHofer** [comment]
Porting your tests requires to Assert the negative ones which throw i.e.:

```csharp
AssertExtensions.Throws<ArgumentException>(ParseFailure.NotEnoughBrackets, () => ...);
```

I want to avoid asserting on strings and I believe passing a FailureReason is a better differentiator overall.

**Comment by CyrusNajmabadi** [comment]
I'm asserting on strings *very* intentionally.  Because i want to know that i'm giving hte user the same error message that you guys are giving.  That's because i want it to seem as if they're actually running your regex parser under the covers.

This is necessary so that the IDE and runtime experience are unified.  i.e. if, at runtime, you gave a very specific error message, then during IDE time, we should give the same error message as well.  If we gave something different, that would be very confusing.

However, **if**:

1. You had a richer error, that had a diagnostic *and* a position in the source string. and
2. you could just parse, without any concerns of OOMing and whatnot (like today)

Then we could consider not having the roslyn parser report any diagnostics at all.  We could just defer to you guys to generate the error and give it's position.  We could then just call into you to analyze the regex and we would map that error location back to the original location in the user's source.

I avoided any approach like that in Roslyn because i found i could not trust the perf characteristics of the regex library. In other words, there was no knowing what sort of memory or CPU overhead might blow up when calling into the library.  This is understandable given that this lib is intended not just for parsing, but for also actually execution at runtime.  But htat did make it unsuitable for IDE purposes where we would want the cost of parsing to only scale linearly with the size of the initial regex string.

**Comment by danmosemsft** [comment]
XmlException is a precedent for line/column on the exception (as properties not text, so editor can jump to the place)

**Comment by CyrusNajmabadi** [comment]
Note: @danmosemsft fixed a bunch of issues i reported.  However:

1. I still don't know if hte regex library is hardened enough to use it, while passing it arbitrary user input.  For example, indexing out of bounds, stack overflows, etc.
2. I don't know if it can still use tons of memory in some scenarios (even though it was hardened for one scenario).  
3. I don't know if it does anything that might consume tons of CPU.

Basically, all things that would be bad in an IDE, but woudl be fine in a program that does not accept user input, or is fine just killing things if there is bad behavior.  

If we could get such hardening constraints from the existing library, then we might be able to use it.

**Comment by CyrusNajmabadi** [comment]
> XmlException is a precedent for line/column on the exception (as properties not text, so editor can jump to the place)

Right.  Though please don't do line/col as no one ever agrees on what they mean :D   Just do "position in string" as that's super well defined, and makes sense given that it's a string that is fed into the regex engine :D

Thanks!
